File: versions/2_0/hack/hack.py
from optimum.intel.openvino import OVModelForFeatureExtraction
from transformers import AutoTokenizer
from openvino.runtime import Core
import numpy as np

from optimum.exporters.onnx import main_export

model_id = "sentence-transformers/paraphrase-MiniLM-L3-v2"
output_path = "./onnx_model"

main_export(
    model_name_or_path=model_id,
    output=output_path,
    task="feature-extraction",
    dtype="fp32",  # ONNX must be FP32 initially
    pad_token_id=0,
    max_length=128
)

# === 1. Export to OpenVINO FP16 (if not already done) ===
model_id = "sentence-transformers/paraphrase-MiniLM-L3-v2"
export_dir = "./openvino_model_fp16"

model = OVModelForFeatureExtraction.from_pretrained(
    model_id,
    export=True,
    compile=False,
    dtype="fp16",
    export_kwargs={
        "input": "input_ids[1,128],attention_mask[1,128]"
    }
)
model.save_pretrained(export_dir)

# === 2. Load Tokenizer ===
tokenizer = AutoTokenizer.from_pretrained(model_id)

# === 3. Tokenize input ===
text = ["The quick brown fox jumps over the lazy dog."]
tokens = tokenizer(text, padding="max_length", truncation=True, max_length=128, return_tensors="np")

# === 4. Run Inference with OpenVINO Runtime on NPU ===
core = Core()
model = core.read_model(f"{export_dir}/openvino_model.xml")
compiled_model = core.compile_model(model, "AUTO")  # try "AUTO" if NPU fails

infer_request = compiled_model.create_infer_request()

# === 5. Prepare Input Tensors ===
input_ids_tensor = tokens["input_ids"]
attention_mask_tensor = tokens["attention_mask"]

infer_request.set_tensor(compiled_model.input("input_ids"), input_ids_tensor)
infer_request.set_tensor(compiled_model.input("attention_mask"), attention_mask_tensor)

# === 6. Inference ===
infer_request.infer()
output_tensor = infer_request.get_tensor(compiled_model.output(0))
output = output_tensor.data  # shape: [1, seq_len, hidden_size]

# === 7. Extract [CLS] Embedding ===
cls_embedding = output[0][0]  # assuming [CLS] is at index 0
print("CLS Embedding (first 5 dims):", cls_embedding[:5])

File: versions/2_0/hack/hack.js
import { AutoTokenizer } from '@huggingface/transformers';
import { addon as ov } from "openvino-node";


const model_name = "sentence-transformers/paraphrase-MiniLM-L3-v2"
try {

    const tokenizer = await AutoTokenizer.from_pretrained(model_name);
    const tokens = await tokenizer(["testing testing what is up"], {
        padding: true,
        truncation: true,
        return_tensor: "np"
    });
    console.log(tokens);

    // Setup
    const core = new ov.Core();
    console.log(core.getAvailableDevices());

    const model = await core.readModel(`./${model_name.replace("/","_")}_openvino_fp16/openvino_model.xml`);

    // console.log(core.queryModel(model, "NPU"));
    console.log("Model loaded successfully");
    console.log(core.getVersions("NPU")) 
    console.log(core.getProperty("NPU","SUPPORTED_PROPERTIES"))
    console.log(core.getProperty("NPU", "FULL_DEVICE_NAME"))
    console.log(core.getProperty("NPU", "CACHE_DIR"))
    console.log(core.getProperty("CPU", "CACHE_DIR"))
    console.log(core.getProperty("NPU", "RANGE_FOR_ASYNC_INFER_REQUESTS"))

    // core.setProperty({ "ENABLE_DEVICE_NPU": "true"});
    const compiled = await core.compileModel(model, "MULTI:NPU,CPU");


    console.log("Model compiled successfully");
    const infer = compiled.createInferRequest();

    // Tokenized input
    // const inputIds = Array.from(tokens.input_ids.data);
    console.log("Input IDs:", tokens.input_ids.data);
    console.log("Attention Mask:", tokens.attention_mask.data);
    // const inputIds = BigInt64Array.from(tokens.input_ids.tolist().flat(), Number);
    // const attentionMask = BigInt64Array.from(tokens.attention_mask.tolist().flat(), Number);

    const flatIds = Array.from(tokens.input_ids.tolist().flat(), Number);
    const flatMask = Array.from(tokens.attention_mask.tolist().flat(), Number);


    // const inputIds = tokens.input_ids.data
    // const attentionMask = tokens.input_ids.data;
    const inputShape = tokens.input_ids.dims;
    console.log("Input shape:", inputShape);
    const batch = inputShape[0];
    const seqLen = inputShape[1];

    console.log(`Batch: ${batch}, Sequence Length: ${seqLen}`)
    console.log("Input IDs:", flatIds);
    console.log("Attention Mask:", flatMask);
    console.log("Input IDs shape:", tokens.input_ids.dims);
    console.log("Attention Mask shape:", tokens.attention_mask.dims);
    console.log("batch:", batch);

    console.log("seqLen:", seqLen);

    const idsArray = new BigInt64Array(flatIds.map(BigInt));       // `flatIds` is a normal JS array of numbers
    const maskArray = new BigInt64Array(flatMask.map(BigInt));     // same for attention mask

    const idsTensor = new ov.Tensor("i64", [batch, seqLen], idsArray);
    const maskTensor = new ov.Tensor("i64", [batch, seqLen], maskArray);

    // Convert to OV tensors
    // const idsTensor = new ov.Tensor("i32", [batch, seqLen], inputIds);
    // const maskTensor = new ov.Tensor("i32", [batch, seqLen], attentionMask);

    // Inference
    await infer.setInputTensor(0, idsTensor);
    await infer.setInputTensor(1, maskTensor);
    const tokenTypeArray = new BigInt64Array(batch * seqLen).fill(0n);
    const tokenTypeTensor = new ov.Tensor("i64", [batch, seqLen], tokenTypeArray);
    infer.setInputTensor(2, tokenTypeTensor);
    await infer.infer();

    // Output tensor
    const output = await infer.getOutputTensor(0); // shape: [batch, seqLen, hiddenSize]
    console.log("Output shape:", output);

    // Mean pooling
    const outData = output.data;
    console.log("Output data:", outData);
    const hiddenSize = output.dims[2];
    const embeddings = [];

    for (let b = 0; b < batch; b++) {
        const emb = new Float32Array(hiddenSize);
        let valid = 0;

        for (let t = 0; t < seqLen; t++) {
            if (attentionMask[b][t] === 0) continue;
            valid++;
            for (let h = 0; h < hiddenSize; h++) {
                const idx = b * seqLen * hiddenSize + t * hiddenSize + h;
                emb[h] += outData[idx];
            }
        }
        embeddings.push(emb.map(v => v / valid));
    }
    console.log("hi")

    console.log({ embeddings });
} catch (error) {
    console.error("Error:", error);
}
